### PR TITLE
Uploaded at column in TB

### PR DESCRIPTION
**Pull request by GustavLagneborgScilifelab**
This PR adds/fixes ...
Adds a uploaded at column in TB when an MIP, BALSAMIC or FLUFFY analysis is uploaded in CG

### How to prepare for test
- [ ] ssh to hasta (depending on type of change)
- [ ] activate stage: `us`
- [ ] request trailblazer-stage on hasta: `paxa`
- [ ] install on stage:
    ```Shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t trailblazer -b uploaded-at-column -a
    ```
- [ ] ssh to clinical-db (depending on type of change)
- [ ] install on stage:
`bash /home/proj/production/servers/resources/clinical-db.scilifelab.se/update-trailblazer-ui-stage.sh [THIS-BRANCH-NAME]`

### How to test
- [ ] login to ...
- [ ] do ...

### Expected test outcome
- [ ] check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review
- [ ] code approved by
- [ ] tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

File: alembic/versions/fca618ecc3ad_.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+"""adds a uploaded at column when an analysis is uploaded to its corresponding place
```
Comment (henrikstranneheim) [inline alembic/versions/fca618ecc3ad_.py:1]:
```suggestion
"""Adds a uploaded at column for showing an analysis finished uploading.
```

File: tests/store/test_store_api.py [hunk 2]
```diff
@@ -101,6 +101,20 @@ def test_is_latest_analysis_ongoing(sample_store, family: str, expected_bool: bo
     assert is_ongoing is expected_bool
 
 
+def test_set_analysis_uploaded(sample_store):
```
Comment (henrikstranneheim) [inline tests/store/test_store_api.py:104]: type hint

File: tests/store/test_store_api.py [hunk 3]
```diff
@@ -101,6 +101,20 @@ def test_is_latest_analysis_ongoing(sample_store, family: str, expected_bool: bo
...
+    """Test setting analysis uploaded at for an analysis."""
+
+    # GIVEN a store with an analysis
+    analysis_obj = sample_store.analyses().first()
+    uploaded_at = datetime.datetime.now()
+
+    # WHEN setting an analysis uploaded at
+    sample_store.set_analysis_uploaded(case_id=analysis_obj.family, uploaded_at=uploaded_at)
```
Comment (henrikstranneheim) [inline tests/store/test_store_api.py:112]:
```suggestion

    # WHEN setting an analysis uploaded at
    sample_store.set_analysis_uploaded(case_id=analysis_obj.family, uploaded_at=timestamp_now)
```


Create timestamp_now fixture. Steal it from cg

File: trailblazer/server/api.py [hunk 4]
```diff
@@ -247,3 +247,16 @@ def post_add_pending_analysis():
         return jsonify(**data), 201
     except Exception as e:
         return jsonify(f"Exception: {e}"), 409
+
+
+@blueprint.route("/set_analysis_uploaded", methods=["PUT"])
+def put_set_analysis_uploaded():
+    content = request.json
```
Comment (henrikstranneheim) [inline trailblazer/server/api.py:254]: type hint

File: trailblazer/server/api.py [hunk 5]
```diff
@@ -247,3 +247,16 @@ def post_add_pending_analysis():
...
+
+    try:
+        store.set_analysis_uploaded(
+            case_id=content.get("case_id"), uploaded_at=content.get("uploaded_at")
+        )
+        return jsonify("Success! Update request sent"), 201
+    except Exception as e:
+        return jsonify(f"Exception: {e}"), 409
```
Comment (henrikstranneheim) [inline trailblazer/server/api.py:262]:
```suggestion
    except Exception as error:
        return jsonify(f"Exception: {error}"), 409
```

File: trailblazer/store/api.py [hunk 6]
```diff
@@ -248,6 +248,13 @@ def set_analysis_completed(self, analysis_id: int) -> None:
         self.commit()
         LOG.info(f"{analysis_obj.family} - status set to COMPLETED")
 
+    def set_analysis_uploaded(self, case_id: str, uploaded_at: dt.datetime) -> None:
+        """Setting analysis uploaded at."""
+        analysis_obj = self.get_latest_analysis(case_id=case_id)
```
Comment (henrikstranneheim) [inline trailblazer/store/api.py:253]: type hint

File: trailblazer/store/api.py [hunk 7]
```diff
@@ -248,6 +248,13 @@ def set_analysis_completed(self, analysis_id: int) -> None:
...
+        analysis_obj.uploaded_at = uploaded_at
```
Comment (henrikstranneheim) [inline trailblazer/store/api.py:254]: type hint

**Review by henrikstranneheim — COMMENTED** [review]
👍

**Comment by GustavLagneborgScilifelab** [comment]
MIP case have an uplaoded-at date when uploaded in CG (click on the image):
Screenshot 2022-12-06 at 15 13 51

BALSAMIC case have an uplaoded-at date when uploaded in CG (click on the image):
Screenshot 2022-12-06 at 15 13 36

Rsync jobs are now submitted with the following family name: <CASE_ID>_rsync:
Screenshot 2022-12-06 at 15 24 49

File: tests/conftest.py [hunk 8]
```diff
@@ -74,3 +75,9 @@ def sample_store(store):
 def trailblazer_context(sample_store):
     """Trailblazer context to be used in CLI"""
     return {"trailblazer": sample_store}
+
+
+@pytest.fixture(scope="function", name="timestamp_now")
```
Comment (henrikstranneheim) [inline tests/conftest.py:80]:
```suggestion
@pytest.fixture(name="timestamp_now")
```

**Review by henrikstranneheim — APPROVED** [review]
👍 . Should you update this branch with master before merging?

**Comment by GustavLagneborgScilifelab** [comment]
Successfully installed in production:
Screenshot 2022-12-08 at 14 40 17

**Pull request merged**